src/cache.jl

```diff
@@ -3,29 +3,82 @@
 using Core.Compiler: retrieve_code_info, CodeInfo, MethodInstance, SSAValue, SlotNumber, ReturnNode
 using Base: _methods_by_ftype
 
-# generated function that crafts a custom code info to call the actual compiler.
-# this gives us the flexibility to insert manual back edges for automatic recompilation.
+# generated function that returns the world age of a compilation job. this can be used to
+# drive compilation, e.g. by using it as a key for a cache, as the age will change when a
+# function or any called function is redefined.
+
+
+"""
+    get_world(ft, tt)
+
+A special function that returns the world age in which the current definition of function
+type `ft`, invoked with argument types `tt`, is defined. This can be used to cache
+compilation results:
+
+    compilation_cache = Dict()
+    function cache_compilation(ft, tt)
+        world = get_world(ft, tt)
+        get!(compilation_cache, (ft, tt, world)) do
+            # compile
+        end
+    end
+
+What makes this function special is that it is a generated function, returning a constant,
+whose result is automatically invalidated when the function `ft` (or any called function) is
+redefined. This makes this query ideally suited for hot code, where you want to avoid a
+costly look-up of the current world age on every invocation.
+
+Normally, you shouldn't have to use this function, as it's used by `FunctionSpec`.
+
+!!! warning
+
+    Due to a bug in Julia, JuliaLang/julia#34962, this function's results are only
+    guaranteed to be correctly invalidated when the target function `ft` is executed or
+    processed by codegen (e.g., by calling `code_llvm`).
+"""
+get_world
+
+# generate functions currently do not know which world they are invoked for, so we fall
+# back to using the current world. this may be wrong when the generator is invoked in a
+# different world (TODO: when does this happen?)
 #
-# we also increment a global specialization counter and pass it along to index the cache.
-
-const specialization_counter = Ref{UInt}(0)
-@generated function specialization_id(job::CompilerJob{<:Any,<:Any,FunctionSpec{f,tt}}) where {f,tt}
-    # get a hold of the method and code info of the kernel function
-    sig = Tuple{f, tt.parameters...}
-    # XXX: instead of typemax(UInt) we should use the world-age of the fspec
-    mthds = _methods_by_ftype(sig, -1, typemax(UInt))
+# XXX: this should be fixed by JuliaLang/julia#48611
+
+function get_world_generator(self, ::Type{Type{ft}}, ::Type{Type{tt}}) where {ft, tt}
```

Comment:
This forces recompilation of this method for every combination of input types, which is probably not what you wanted here. These should probably just be normal arguments:
```
using Core.Compiler: isType
function get_world_generator(self, ft::Type, tt::Type)
    @assert isType(ft) && isType(tt) # as was declared in the signature
    ft = ft.parameters[1]
    tt = tt.parameters[1]
```